src/rust/engine/fs/src/directory.rs

```diff
@@ -499,29 +548,43 @@ impl DigestTrie {
         Entry::File(f) => {
           digests.push(f.digest);
         }
+        // There is no digest for a symlink
+        Entry::Symlink(_) => (),
       }
     }
     digests
   }
 
-  /// Return a pair of Vecs of the file paths and directory paths in this DigestTrie, each in
-  /// sorted order.
-  ///
-  /// TODO: This should probably be implemented directly by consumers via `walk`, since they
-  /// can directly allocate the collections that they need.
-  pub fn files_and_directories(&self) -> (Vec<PathBuf>, Vec<PathBuf>) {
```

Comment:
I split this into 3 implementations, as only one caller needed all the items at the same time, and it was for `__repr__`, so perf doesn't matter __that__ much.